libImaging/Storage.c

```diff
@@ -265,46 +264,184 @@ ImagingDelete(Imaging im)
+        }
+    } else {
+        if (dirty) {
+            block.ptr = malloc(requested_size);
+        } else {
+            block.ptr = calloc(1, requested_size);
+        }
+        arena->stats_allocated_blocks += 1;
+    }
+    block.size = requested_size;
+    return block;
+}
+
+void
+memory_return_block(ImagingMemoryArena arena, ImagingMemoryBlock block)
+{
+    if (arena->blocks_cached < arena->blocks_max)  {
+        // Reduce block size
+        if (block.size > arena->block_size) {
+            block.size = arena->block_size;
+            block.ptr = realloc(block.ptr, arena->block_size);
+        }
+        arena->blocks[arena->blocks_cached] = block;
+        arena->blocks_cached += 1;
+    } else {
+        free(block.ptr);
+        arena->stats_freed_blocks += 1;
+    }
+}
+
+
 static void
 ImagingDestroyArray(Imaging im)
 {
-    int y;
+    int y = 0;
 
-    if (im->image)
-        for (y = 0; y < im->ysize; y++)
-            if (im->image[y])
-                free(im->image[y]);
+    if (im->blocks) {
+        while (im->blocks[y].ptr) {
+            memory_return_block(&ImagingDefaultArena,  im->blocks[y]);
+            y += 1;
+        }
+        free(im->blocks);
+    }
 }
 
 Imaging
-ImagingAllocateArray(Imaging im, int dirty)
+ImagingAllocateArray(Imaging im, int dirty, int block_size)
 {
-    ImagingSectionCookie cookie;
-
-    int y;
-    char* p;
+    int y, line_in_block, current_block;
+    ImagingMemoryArena arena = &ImagingDefaultArena;
+    ImagingMemoryBlock block = {NULL, 0};
+    int aligned_linesize, lines_per_block, blocks_count;
+    char *aligned_ptr;
+
+    /* 0-width or 0-height image. No need to do anything */
+    if ( ! im->linesize || ! im->ysize) {
+        return im;
+    }
 
-    ImagingSectionEnter(&cookie);
+    aligned_linesize = (im->linesize + arena->alignment - 1) & -arena->alignment;
+    lines_per_block = (block_size - (arena->alignment - 1)) / aligned_linesize;
+    if (lines_per_block == 0)
+        lines_per_block = 1;
+    blocks_count = (im->ysize + lines_per_block - 1) / lines_per_block;
+    // printf("NEW size: %dx%d, ls: %d, lpb: %d, blocks: %d\n",
+    //        im->xsize, im->ysize, aligned_linesize, lines_per_block, blocks_count);
+
+    /* One extra ponter is always NULL */
+    im->blocks = calloc(sizeof(*im->blocks), blocks_count + 1);
+    if ( ! im->blocks) {
+        return (Imaging) ImagingError_MemoryError();
+    }
 
     /* Allocate image as an array of lines */
+    line_in_block = 0;
+    current_block = 0;
     for (y = 0; y < im->ysize; y++) {
-        /* malloc check linesize checked in prologue */
-        if (dirty) {
-            p = (char *) malloc(im->linesize);
-        } else {
-            p = (char *) calloc(1, im->linesize);
-        }
-        if (!p) {
-            ImagingDestroyArray(im);
-            break;
+        if (line_in_block == 0) {
+            int required;
+            int lines_remained = lines_per_block;
```

Comment:
I think this is better named `lines_remaining`

Reply:
Done